### DISCUSSION TITLE
【求助/讨论】卡牌权重设置的通用规则

### DISCUSSION THREAD

**Discussion by John-bai-Smith** [xjw580/Hearthstone-Script#184] [General]
发现4.2更新了卡牌权重设置，喜大普奔啊！
想要尝试一下，仔细一想却感觉无从下手。
之前用激进策略挂机的时候，观测到一个问题，打牌只会优先从高费卡开始打 ，减到0费的卡当回合识别不到
这就会导致明明手里有1费随从，结果高费把奇利亚斯打了的情况，这种情况能够通过权重设置解决吗？

**Comment by xjw580** [comment]
目前权重只影响出哪些牌，决定了出哪些牌后出牌的顺序不影响，下版本可加一个出牌顺序权重设置

**Comment by xjw580** [comment]
v4.2.1-TEST 版本已加入出牌权重

**Comment by LuoShen1512** [comment]
我想问下是怎么样一个逻辑 1、每张卡牌的初始权重是一样的吗 ？  2、我在权重设置里面将某张卡牌的权重增加后，是会优先打出这张卡牌吗，如果可以，我要把权重设置成多少才能达到优先打出的效果？

**Comment by xjw580** [comment]
> 我想问下是怎么样一个逻辑 1、每张卡牌的初始权重是一样的吗 ？ 2、我在权重设置里面将某张卡牌的权重增加后，是会优先打出这张卡牌吗，如果可以，我要把权重设置成多少才能达到优先打出的效果？

这个权重只是个参考，要看插件的策略根不根据这个计算了，自带的策略首先从设置里的权重里读，读不到则默认为1，然后根据卡牌是否有亡语，圣盾等再加减，下面是解场的权重计算，出牌的计算自行查阅代码
https://github.com/xjw580/Hearthstone-Script/blob/6e91b2b5efe4b9187d55c56c899788786347d1f0/Hearthstone-Script-Base-Deck/src/main/kotlin/club/xiaojiawei/util/DeckStrategyUtil.kt#L131-L159

**Comment by LuoShen1512** [comment]
谢谢  加减范围一般都是0.3的话  
那我设置成10一般是包出了

**Comment by DaveGitLi** [comment]
@xjw580 设置权重的区间建议是多少呢， DeckStrategyUtil.cleanPlay()这个方法执行随从攻击的逻辑在哪里呀，我没有找到0.0

**Comment by xjw580** [comment]
> @xjw580 设置权重的区间建议是多少呢， DeckStrategyUtil.cleanPlay()这个方法执行随从攻击的逻辑在哪里呀，我没有找到0.0

所有调用了execAction方法的地方
https://github.com/xjw580/Hearthstone-Script/blob/6e91b2b5efe4b9187d55c56c899788786347d1f0/Hearthstone-Script-Base-Deck/src/main/kotlin/club/xiaojiawei/util/DeckStrategyUtil.kt#L226-L227

**Comment by DaveGitLi** [comment]
我看这两个execAction是计算死亡的卡片数量。
calcClean是通过反演逻辑，并且通过异步计算提升效率，最终返回一个计算结果。没有找到随从攻击具体那张卡的步骤，我是想向攻击步骤内，加一些法术和技能的判断

**Comment by xjw580** [comment]
> 我看这两个execAction是计算死亡的卡片数量。 calcClean是通过反演逻辑，并且通过异步计算提升效率，最终返回一个计算结果。没有找到随从攻击具体那张卡的步骤，我是想向攻击步骤内，加一些法术和技能的判断

攻击行为在attack方法
https://github.com/xjw580/Hearthstone-Script/blob/6e91b2b5efe4b9187d55c56c899788786347d1f0/Hearthstone-Script-Base-Deck/src/main/kotlin/club/xiaojiawei/util/DeckStrategyUtil.kt#L486-L497
